### DISCUSSION TITLE
How to handle/type errors?

### DISCUSSION THREAD

**Discussion by Falci** [wevm/wagmi#2101] [Q&A]
What's the type of an error from `const {error} = usePrepareContractWrite({...});`?

Typescript says it's just the native `Error`, but I see it has more properties:

```json
{
    "reason": "execution reverted: domain already registered",
    "code": "UNPREDICTABLE_GAS_LIMIT",
    "method": "estimateGas",
    "transaction": {
      "from": "0xb33587...",
      "to": "0x2dfd9a...",
      "value": {
        "type": "BigNumber",
        "hex": "0x0d0e64..."
      },
      "data": "0xc8d26a...",
      "accessList": null
    },
    "error": {
      "code": 3,
      "data": "0x08c379...",
      "message": "execution reverted: domain already registered"
    }
  }
```

I tried to cast it to `ContractMethodRevertedError` but doesn't seems to be the same kind neither:
<img width="725" alt="image" src="https://user-images.githubusercontent.com/1330573/228818055-6294a85d-00a6-4a54-895a-68d9163d35b1.png">

Without the proper cast I can't safely access the property:
<img width="490" alt="image" src="https://user-images.githubusercontent.com/1330573/228818555-5cad3b44-edb1-48d8-972f-31b0eb787ecf.png">

**Comment by Falci** [comment]
Related:
https://github.com/wagmi-dev/wagmi/discussions/233

https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/logger/lib/index.js#L237-L244

**Reply by tmm** [reply]
Errors will be improved when wagmi migrates from ethers to [viem](https://viem.sh). Right now, ethers will throw a variety of errors internally that aren't standardized so you'll need to be on the lookout for those. wagmi [exports some standardized errors](https://github.com/wagmi-dev/wagmi/blob/main/packages/react/src/index.ts#L54-L71) that might be useful in the meantime.

https://twitter.com/jakemoxey/status/1633240447761121281
